Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -92,55 +92,70 @@ class ThreadSafeWeakHashSet final {
 
     const_iterator end() const { return { { } }; }
 
-    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
-    void add(const U& value)
+    template<typename U>
+    void add(const U& value) requires (std::is_convertible_v<U*, T*>)
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!value.controlBlock().objectHasStartedDeletion());
         Locker locker { m_lock };
         ControlBlockRefPtr retainedControlBlock { &value.controlBlock() };
-        if (!retainedControlBlock)
-            return;
+        ASSERT(retainedControlBlock);
         amortizedCleanupIfNeeded();
-        m_map.add(static_cast<const T*>(&value), WTFMove(retainedControlBlock));
+        m_set.add(std::make_pair(WTFMove(retainedControlBlock), &value));
     }
 
-    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
-    bool remove(const U& value)
+    template<typename U>
+    bool remove(const U& value) requires (std::is_convertible_v<U*, T*>)
     {
         Locker locker { m_lock };
         amortizedCleanupIfNeeded();
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
+        if (!value.weakRefCount())
```

Comment:
this seems to be potentially racy if a weakRef is taken between the time it's removed.
The previous code avoided this situation completely.

Reply:
@jyavenard  and I had a discussion about this offline. There's not actually a race here since the presence of the value's controlBlock in the set implies it must have a weakRefCount of at least 1 (the set's ref). Thus the contrapositive says a weakRefCount of less than 1 (aka 0) means it can't be in the set.